python/ray/dag/tests/experimental/test_mocked_nccl_dag.py

```diff
@@ -0,0 +1,355 @@
+# coding: utf-8
+import os
+import sys
+import torch
+
+import pytest
+
+import ray
+import ray.cluster_utils
+from ray.experimental.channel.torch_tensor_type import TorchTensorType
+from ray.experimental.channel.conftest import (
+    Barrier,
+    start_nccl_mock,
+)
+from ray.tests.conftest import *  # noqa
+from ray.dag import InputNode
+
+
+@ray.remote(num_cpus=0, num_gpus=1)
+class MockedWorker:
+    def __init__(self):
+        self.chan = None
+
+    def start_mock(self):
+        """
+        Patch methods that require CUDA.
+        """
+        start_nccl_mock()
+
+    def send(self, shape, dtype, value: int, send_as_dict=False):
+        if send_as_dict:
+            return self.send_dict([(value, value, shape, dtype)])
+
+        return torch.ones(shape, dtype=dtype) * value
+
+    def recv(self, tensor):
+        if isinstance(tensor, dict):
+            assert len(tensor) == 1
+            tensor = tensor.values()[0]
+
+        return (tensor[0].item(), tensor.shape, tensor.dtype)
+
+    def send_dict(self, entries):
+        results = {}
+        for key, value, shape, dtype in entries:
+            results[key] = torch.ones(shape, dtype=dtype) * value
+        return results
+
+    def recv_dict(self, tensor_dict):
+        results = []
+        for key in sorted(tensor_dict.keys()):
+            tensor = tensor_dict[key]
+            results.append((key, tensor[0].item(), tensor.shape, tensor.dtype))
+        return results
+
+
+@pytest.mark.parametrize(
+    "ray_start_cluster",
+    [
+        {
+            "num_cpus": 2,
+            "num_gpus": 2,
+            "num_nodes": 1,
+        }
+    ],
+    indirect=True,
+)
+def test_p2p(ray_start_cluster):
+    """
+    Test simple sender -> receiver pattern. Check that receiver receives
+    correct results.
+    """
+    # Barrier name should be barrier-{sender rank}-{receiver rank}.
+    # Create a barrier in both directions because we don't know which rank will
+    # get assigned to sender and receiver.
+    barrier1 = Barrier.options(name="barrier-0-1").remote()  # noqa
+    barrier2 = Barrier.options(name="barrier-1-0").remote()  # noqa
+
+    sender = MockedWorker.remote()
+    receiver = MockedWorker.remote()
+
+    ray.get([sender.start_mock.remote(), receiver.start_mock.remote()])
+
+    shape = (10,)
+    dtype = torch.float16
+
+    # Test torch.Tensor sent between actors.
+    with InputNode() as inp:
+        dag = sender.send.bind(
+            inp.shape, inp.dtype, inp[0], send_as_dict=inp.send_as_dict
```

Comment:
One question about the output schema, what type of outputs format does our auto infer feature support?IIUC, the actor method can only return one value, the type should be either "a single tensor" or "a dict of tensors"? 

Does that contradict with [this PR](https://github.com/ray-project/ray/pull/47024)? where ADAG can create multiple channels for multiple return values

Reply:
Great question.

The type can be anything, although types with simple serialization methods are best since this PR requires the user to guarantee serialization order of tensors.

This PR should be compatible with #47024. #47024 works as long as:
1. the actor method is annotated with `num_returns=N`
2. the actor method returns a tuple of N results.
If (2) does not match (1), then an error is thrown.

This PR applies to individual results. So for this kind of function:

```python
def return_two(self):
  return torch.tensor(...), {
    "a": torch.tensor(...),
    "b": torch.tensor(...),
  }
```

if there is no `num_returns` annotation, then the actor returns one value and the schema would be `Tuple[Tensor, Dict[str, Tensor]]`. If `num_returns=2`, then the actor will send on two different channels, one with schema `Tensor` and the other with schema `Dict[str, Tensor]`.